tests/e2e/tests/pilot/performance/serviceentry_test.go

```diff
@@ -0,0 +1,346 @@
+
+	test := newTestParam(*serviceEntryCount, *configRefreshInterval, *terminateAfter)
+	go test.init()
+
+	t.Log("building & starting mock mcp server...")
+	mcpServer, err := runMcpServer(test)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	defer mcpServer.Close()
+
+	tearDown := initLocalPilotTestEnv(t, mcpServer.Port, pilotGrpcPort, pilotDebugPort)
+	defer tearDown()
+
+	t.Log("run edge router envoy...")
+	gateway := runEnvoy(t, pilotGrpcPort, pilotDebugPort)
+	defer gateway.TearDown()
+
+	t.Log("check for service entries in debug endpoint...")
+	g.Eventually(func() (int, error) {
+		return registeredServiceEntries(fmt.Sprintf("http://127.0.0.1:%d/debug/configz", pilotDebugPort))
+	}, "30s", "1s").Should(gomega.Equal(test.configCount))
+
+	test.terminate()
+}
+
+type testParam struct {
+	configCount       int
+	interval          time.Duration
+	terminateAfter    time.Duration
+	cachedResponseMux sync.RWMutex
+	cachedResponse    []*mcp.Envelope
+	quit              chan struct{}
+}
+
+func newTestParam(count, interval, after int) *testParam {
+	t := &testParam{
+		configCount:    count,
+		interval:       time.Duration(interval),
+		terminateAfter: time.Duration(after),
+		quit:           make(chan struct{}),
+	}
+	t.generateServiceEntries()
+	return t
+}
+
+func (t *testParam) init() {
+	// this is to create a slightly more realistic scenario by not
+	// overwhelming the client with new service entries on every watch call
+	ticker := time.NewTicker(time.Second * t.interval)
+	for {
+		select {
+		case <-ticker.C:
+			// update cache
+			t.generateServiceEntries()
+		case <-t.quit:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+func (t *testParam) terminate() {
+	<-time.After(time.Second * t.terminateAfter)
+	t.quit <- struct{}{}
+}
+
+func (t *testParam) mcpServerResponse(req *mcp.MeshConfigRequest) (*mcpserver.WatchResponse, mcpserver.CancelWatchFunc) {
+	var cancelFunc mcpserver.CancelWatchFunc
+	cancelFunc = func() {
+		log.Printf("watch canceled for %s\n", req.GetTypeUrl())
+	}
+	// registering a wildcard gateway so that we can at least test some of the service entries
+	// to see if they are registered
+	if req.GetTypeUrl() == fmt.Sprintf("type.googleapis.com/%s", model.Gateway.MessageName) {
+		return generateGateway(req), cancelFunc
+	}
+	if req.GetTypeUrl() == fmt.Sprintf("type.googleapis.com/%s", model.ServiceEntry.MessageName) {
+		return &mcpserver.WatchResponse{
+			Version:   req.GetVersionInfo(),
+			TypeURL:   req.GetTypeUrl(),
+			Envelopes: t.cachedResponses(),
+		}, cancelFunc
+	}
+
+	return &mcpserver.WatchResponse{
+		Version:   req.GetVersionInfo(),
+		TypeURL:   req.GetTypeUrl(),
+		Envelopes: []*mcp.Envelope{},
+	}, cancelFunc
+}
+
+func (t *testParam) cachedResponses() []*mcp.Envelope {
+	t.cachedResponseMux.RLock()
+	defer t.cachedResponseMux.RUnlock()
+	return t.cachedResponse
+}
+
+func (t *testParam) generateServiceEntries() (envelopes []*mcp.Envelope) {
+	servicePorts := availablePorts(t.configCount)
```

Comment:
Do we need to use real (i.e. unused) ports here if we switch to adsc.go and the pilot load generates are synthetic?